Feature/cli timestamp: Allows the user to have timestamps in the CLI report

jimc404 opened the pull request:
Fixes #1998

### What does this PR do?
I have implemented two flags in this PR:

1. ```--reporter-cli-show-timestamp```: This appends a timestamp to each request displaying their start time. It causes the following output:
```
→ A simple GET request
  GET https://postman-echo.com/get?source=newman-sample-github-collection [200 OK, 637B, 1356ms]
```
turn to
```
→ A simple GET request
  Mar 02 2020 03:37:39 GMT+0530  GET https://postman-echo.com/get?source=newman-sample-github-collection [200 OK, 637B, 1520ms]
```
This timestamp also displays the UTC offset of the timezone configured on the system.

The timezone info may not be desired and thus to allow for logging of simpler timestamps there's this flag:
2. ```--reporter-cli-show-timestamp-no-tz```:  This only prints the date and time of the request. The above example with this flag will be:
```
→ A simple GET request
  Mar 02 2020 03:45:56  GET https://postman-echo.com/get?source=newman-sample-github-collection [200 OK, 639B, 1508ms]
```

### Implementation:

- In ```PostmanCLIReporter``` on the event ``` beforeRequest``` check if timestamp options are present.
- If they are present then a string with current system time is fetched. Extract the substring according to the option passed and print it to CLI along with the other data.  

### Testing: 
- For testing of these proposed changes there are 2 new tests added under the file ```test/cli/cli-reporter-show-timestamp.test.js```. 
- They read the CLI output of these features and validate them with regex matches.

jimc404 commented:
@codenirvana Can you provide feedback for this.
TIA

codenirvana commented on lib/reporters/cli/cli-utils.js:
```diff
@@ -75,6 +75,17 @@ cliUtils = {
+     * @returns {String} - A string carrying timestamp
+     */
+    currentTime: function () {
+        return new Date().toString();
```
I feel [toLocaleString](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleString) works best here.

Check if there's any other helpful instance method available in the Date's [prototype](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date).

codenirvana replied:
We can move this method to the CLI reporter itself if it's not used in multiple places.
Better have fewer utils. 😅

jimc404 replied:
> I feel [toLocaleString](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleString) works best here.
> 
> Check if there's any other helpful instance method available in the Date's [prototype](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date).

So, can we use `toUTCString` and `toLocaleString`. I am of the opinion to allow an option to show UTC time, so that users with newman instances in different timezones can use it out of the box. The flags can be `--reporter-cli-show-timestamp` and `--reporter-cli-show-timestamp-utc` or something similar but maybe shorter. 

What do you think?

codenirvana commented on lib/reporters/cli/index.js:
```diff
@@ -57,6 +58,9 @@ extractSNR = function (executions) {
  * @param {Boolean=} reporterOptions.noConsole - Boolean flag to turn off console logging, if set to true.
  * @param {Boolean=} reporterOptions.noBanner - Boolean flag to turn off newman banner, if set to true.
+ * @param {Boolean=} reporterOptions.showTimestamp - Boolean flag to log timestamp of each request, if set to true.
+ * @param {Boolean=} reporterOptions.showTimestampNoTz - Boolean flag to log timestamp of each request, if set to true.
```
no sure if this is still needed.

jimc404 replied:
I added this to maintain consistency with the rest of the JSDoc comment.

jimc404 replied:
Although being the CLI reporter maybe it's okay to not have it. I'm open to both options as it does not affect the usage either way.

codenirvana replied:
I meant having multiple options/formats for logging timestamps.

jimc404 commented:
> I feel [toLocaleString](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleString) works best here.
> 
> Check if there's any other helpful instance method available in the Date's [prototype](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date).

So, can we use `toUTCString` and `toLocaleString`. I am of the opinion to allow an option to show UTC time, so that users with newman instances in different timezones can use it out of the box. The flags can be `--reporter-cli-show-timestamp` and `--reporter-cli-show-timestamp-utc` or something similar but maybe shorter. 

What do you think? @codenirvana

jimc404 commented:
@codenirvana I made the changes you suggested
- `toLocaleString()` is used to show local time.
- Move the one liner function directly into the timestamp code block
- Removed the 2 JsDoc comments
- Updated README

Please review this whenever you get a chance

codenirvana commented on README.md:
```diff
@@ -425,6 +425,9 @@ such a scenario.
 | `--reporter-cli-silent`         | The CLI reporter is internally disabled and you see no output to terminal. |
+
+| `--reporter-cli-show-timestamp` | This prints the local time for each request made. | 
+| `--reporter-cli-show-timestamp-utc` | This prints the UTC time of each request made. |
```
I don't feel any need for this option right now. We can just have a single flag to show the timestamps.

jimc404 replied:
Alright, I'll fix this

codenirvana commented on README.md:
```diff
@@ -425,6 +425,9 @@ such a scenario.
 |-------------|-------------------|
 | `--reporter-cli-silent`         | The CLI reporter is internally disabled and you see no output to terminal. |
+
+| `--reporter-cli-show-timestamp` | This prints the local time for each request made. | 
```
shouldn't this be called `--reporter-cli-show-timestamps` 🤔 , thoughts?
Or, can this change be just part of existing `--verbose` flag?

jimc404 replied:
1. I couldn't decide between choosing `timestamp` and `timestamps`. I went with the former, but since you pointed it out I guess it feels out of place for someone reading it the first time. I'll change it.


2. Since this is a CLI reporter specific option I feel `--timestamps` may be a bit confusing as other flags with `--` work on all reporters.

codenirvana commented on lib/reporters/cli/index.js:
```diff
@@ -147,7 +148,22 @@ PostmanCLIReporter = function (emitter, reporterOptions, options) {
     emitter.on('beforeRequest', function (err, o) {
```
isn't timestamp needed for other events like the script, console, etc?

jimc404 replied:
I implemented what the PR asked for, though adding timestamp for all CLI reported events does seem to have advantages (like checking out what part of the collection is slowing things down). 

It can be added but should I add it in the current flag or will it be okay to add it in the current flag. I'd like some input on this :)

codenirvana commented on lib/reporters/cli/index.js:
```diff
@@ -147,7 +148,22 @@ PostmanCLIReporter = function (emitter, reporterOptions, options) {
+            timestamp = reporterOptions.showTimestampUtc ?
+                currentTime.toUTCString() :
+                currentTime.toLocaleString();
+            print('  %s  %s %s ',
```
should timestamp be wrapped in `[` brackets `]`?

jimc404 replied:
Alright, I'll add brackets and see how it looks. Will commit it before the next review.

jimc404 commented:
@codenirvana I made these changes in the latest push:
- Removed the `utc` timestamp flag
- Renamed the flag to `-show-timestamps`
- Timestamp is now shown in square brackets
Example.
```
→ https://reqres.in/api/users?page=2
  [3/23/2020, 12:39:58 AM]  GET https://reqres.in/api/users/2 [200 OK, 1.04KB, 4.3s]
```
- Added timestamps for `console` and `assertion` events
    - Timestamps could have been added for `beforeItem`, `beforeIteration` and `test` events but they by themselves completed instantly not to mention the events before and after them already had timestamps, so I skipped them. `script` event is already covered by `assertion` event so it wasn't needed. 
    - Date isn't shown for these events (only the time is). This is because the date information was becoming too redundant, date is only shown in the `beforeRequest` event

Let me know if anything else can be fixed

elit-altum commented:
@DannyDainton is this feature request gone stale or can I work on it?
